### PR TITLE
Adding attributes to Params and Fields #521

### PR DESCRIPTION
**Pull request by ambiguousname**
Adds attributes for `Param`, `SelfParam` and `StructField` types.

Not sure where the compiler is getting some of these attribute errors from, but it at least compiles the sample demo just fine. I also didn't set any attributes to inherit, since I don't believe parameters or structure fields have anything that we really want them to inherit from. 

I'll see what I can do to resolve all these attribute errors tomorrow, probably.

Anyways, feedback is welcome!

### CODE REVIEW

File: core/src/hir/lowering.rs [hunk 1]
```diff
@@ -989,11 +991,13 @@ impl<'ast> LoweringContext<'ast> {
                             true,
                         );
 
+                        let attrs = self.attr_validator.attr_from_ast(&self_param.attrs, &Attrs::default(), &mut self.errors);
```
Comment (Manishearth) [inline core/src/hir/lowering.rs:994]: thought: we should add an `attrs.allowed_on_param_or_field()` method to Attrs and check it here and below. I don't think it's a good idea to let people think they can rename fields/params and then have it not work

(eventually i'd like to support param/field renaming though)

**Review by Manishearth — APPROVED** [review]
r+ with a suggestion

File: core/src/hir/attrs.rs [hunk 2]
```diff
@@ -158,11 +155,15 @@ fn maybe_error_unsupported(
...
 impl Attrs {
+    /// `location_validator` represents a callback passed in by `lowering.rs` methods to check if any attributes of a given location are allowed.
+    ///
+    /// Right now this is just used for [`Attrs::allowed_on_param_or_field`], but could be used for other attributes in specific locations.
     pub fn from_ast(
         ast: &ast::Attrs,
         validator: &(impl AttributeValidator + ?Sized),
         parent_attrs: &Attrs,
         errors: &mut ErrorStore,
+        location_validator: Option<fn(syn::Ident) -> Result<(), LoweringError>>,
```
Comment (Manishearth) [inline core/src/hir/attrs.rs:166]: issue: let's not do it this way, this feels overly complicated


instead can we add Param and Field to AttributeContext, and perform this validation during attribute validation?

Comment (Manishearth) [reply core/src/hir/attrs.rs:166]: there's also a good argument to be made for `from_ast` to have its own simpler AttributeContext enum, but it's probably easier to perform further validation in the preexisting validation pass

Comment (ambiguousname) [reply core/src/hir/attrs.rs:166]: Sure, will adjust now

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
sweet!

**Pull request merged**